### PR TITLE
cassandra: reorganize packages and improvements - call for further help

### PR DESCRIPTION
**Pull request by markus1189**
###### Things done:
- [x] Tested using sandboxing (`nix-build --option build-use-chroot true`
- [x] Built on platform: NixOS
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
###### Motivation

I wanted to use cassandra from my NixOS pc, unfortunately I found that the current package does not work:

```
$ nix-env -i cassandra
$ cassandra -f
ERROR 19:58:45 Directory /nix/store/zg11dxs5jjzab6wn3f2b8m56xz4ql88j-cassandra-2.1.13/bin/../data/data doesn't exist
ERROR 19:58:45 Has no permission to create /nix/store/zg11dxs5jjzab6wn3f2b8m56xz4ql88j-cassandra-2.1.13/bin/../data/data directory
```
###### Current State
- I found three `*.nix` files for version 1.2, 2.0 and 2.1 which I refactored into a function `common` in `default.nix`.
- removed unsupported cassandra versions 1.2 and 2.0.  Added 2.2 and 3.0 which are the currently maintained ones
###### TODO

I still have some issues I could not yet resolve.  I hope that somebody here can help me with the following:
- [ ] the `cqlsh` (python script `cqlsh.py`) does not find the `readline` library for python and so there is no tab completion :(  Adding `readline` to the `buildInputs` does not work, I guess I have to somehow wrap the `cqlsh` executable such that it can find it, but I do not yet know how.
- [ ] cassandra needs a `data/data` directory to store data.  Currently it points to the install directory in `/nix/store/...` which is read-only and therefore we cannot even start cassandra successfully (see error at the top).  What is the best place to point cassandra to?  Are there any things to consider given that we might want to support use of different cassandra versions from `nix-shell`?
- [ ] cassandra does it's configuration inside the `conf/` folder, which contains files like `cassandra.yaml` and `logback.xml` which users have to be able to adjust to their liking.  Where do we put those? Again under the aspect that we might want to have multiple versions in use with `nix-shell`.

### CODE REVIEW

File: pkgs/servers/nosql/cassandra/default.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      maintainers = with maintainers; [ nckx rushmorem markus1189 ];
+    };
+  });
+
+in {
+
+ cassandra_2_1 = common {
+    version = "2.1.13";
+    sha256 = "09b3vf5jsv70xlfimj30v8l1zw7c5xdgpw5fpmn6jh8n3gigybqh";
+  };
```
Comment (grahamc) [inline pkgs/servers/nosql/cassandra/default.nix]: Looks like 2.1 is unsupported. Is it? Should we remove it?

Comment (markus1189) [inline pkgs/servers/nosql/cassandra/default.nix]: According to https://cassandra.apache.org/download/: 

> Apache Cassandra 2.1 is supported until November 2016 with critical fixes only. The latest release is 2.1.13, released on 2016-02-08.

But from my side we could also remove 2.1.13

**Comment by copumpkin** [comment]
> cassandra needs a data/data directory to store data. Currently it points to the install directory in /nix/store/... which is read-only and therefore we cannot even start cassandra successfully (see error at the top). What is the best place to point cassandra to? Are there any things to consider given that we might want to support use of different cassandra versions from nix-shell?

I wouldn't default it to anything, if you can avoid it, but see next point.

> cassandra does it's configuration inside the conf/ folder, which contains files like cassandra.yaml and logback.xml which users have to be able to adjust to their liking. Where do we put those? Again under the aspect that we might want to have multiple versions in use with nix-shell.

Is it possible to tell it where to look for cassandra.yaml on the command line or with java/environment variables? The way I'd usually expect this sort of thing to work:

1. Basic nixpkgs package doesn't default to much (or if it does, defaults to storing state in /var and looking for config in /etc, but I'd avoid that if I could)
2. NixOS service module depends on nixpkgs package and constructs appropriately structured cassandra.yaml (and any other config) on the fly when building the system configuration
3. The NixOS module creates a systemd service that starts the cassandra executable, passing it the custom-built configuration

If need be, you can also have the module populate files in /etc for you, but I consider that less elegant (since it relies on what's effectively a global variable being mutated carefully) than just passing in the desired configuration at startup.

**Comment by markus1189** [comment]
> Is it possible to tell it where to look for cassandra.yaml on the command line or with java/environment variables?

Yes we can use `CASSANDRA_CONF`:
>  CASSANDRA_CONF -- Directory containing Cassandra configuration files.

I also think that the best way would be to construct the config files from nix configuration.  I will have a look at writing a service for the package.  

Can anybody help with the missing `readline` support for the `cqlsh` python script?  I already tried adding `wrapProgram $out/bin/cqlsh --prefix PATH : ${python}/bin --prefix PYTHONPATH : "$(toPythonPath ${python.modules.readline})"` but that didn't help either...

**Comment by joachifm** [comment]
Perhaps @FRidh can help you with the python issues

**Comment by FRidh** [comment]
Sorry for the late reply.
Please see https://github.com/markus1189/nixpkgs/pull/1

**Comment by FRidh** [comment]
If I recall correctly @markus1189 and I weren't able to fix this. I am closing this now because it has been dormant now for quite some time.

**Pull request closed without merging**